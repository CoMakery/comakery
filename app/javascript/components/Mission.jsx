import React from 'react'
import PropTypes from 'prop-types'
import { FormLabel } from './subcomponents/FormLabel'
import Alert from './subcomponents/Alert'

export default class Mission extends React.Component {
  constructor(props) {
    super(props)
    this.state = {
      ...props,
      logo       : null,
      image      : null,
      message    : null, // notify sucess or error after account info update
      messageType: 'notice',
      showMessage: false, // show or hide message
      errors     : {}, // error hash for account form
    }
    this.mounted = false
  }

  componentDidMount() {
    this.mounted = true
  }

  componentWillMount() {
    this.mounted = false
  }

  handleChangeFormData = e => {
    const target = e.target
    const value = target.type === 'checkbox' ? target.checked : target.value
    const name = target.name
    console.log(value)
    this.setState({
      [name]: value,
    })
  };

  handleImageChange = e => {
    e.preventDefault()

    let reader = new FileReader()
    let file = e.target.files[0]
    const name = e.target.name

    reader.onloadend = () => {
      this.setState({
        [name]            : file,
        [name + 'Preview']: reader.result
      })
    }

    reader.readAsDataURL(file)
  }

  handleSaveMission = e => {
    e.preventDefault()
    let formData = new FormData()
    const {name, subtitle, description, logo, image} = this.state
    formData.append('mission[name]', name)
    formData.append('mission[subtitle]', subtitle)
    formData.append('mission[description]', description)
    if (logo) {
      formData.append('mission[logo]', logo)
    }
    if (image) {
      formData.append('mission[image]', image)
    }

    // check if it's update or create
    let url = '/missions'
    if (this.props.id) {
      url = '/missions/' + this.props.id
    }

    $.ajax({
      url        : url,
      data       : formData,
      processData: false,
      contentType: false,
      dataType   : 'json',
      type       : this.props.id ? 'PATCH' : 'POST',

      success: response => {
        console.log('success')
      },

      error: (xhr) => {
        let response = xhr.responseJSON
        this.setState({
          message    : response.message,
          messageType: 'alert',
          showMessage: true,
          errors     : response.errors,
        })
        console.log('error')
        console.log(xhr.responseJSON)
      },
    })
  }

  render() {
    console.log(this.state)
    const {id, name, subtitle, description, logoPreview, imagePreview, message, messageType, showMessage, errors} = this.state
    return <React.Fragment>
      <Alert message={message} messageType={messageType} isVisible={showMessage} toggleVisible={() => {
        this.setState({ showMessage: !showMessage })
      }} />

      <form onSubmit={this.handleSaveMission}>
        <div className="mission-container">
          <div className="mission-header">{id ? 'Edit Mission' : 'Create a New Mission'}</div>
          <div className="row">
            <div className="columns large-6 medium-12">
              <FormLabel title="Name" required length={100} currentLength={name.length} />
              <input className="mission-input" type="text" name="name" placeholder="Unicoin" maxLength="100" value={name} onChange={this.handleChangeFormData} />
              {errors.name && <small className="error">{errors.name}</small>}

              <FormLabel title="Subtitle" required length={140} currentLength={subtitle.length} />
              <input className="mission-input" type="text" name="subtitle" placeholder="Unicoin is the future of tokenizing fantastical beast networks" maxLength="140" value={subtitle} onChange={this.handleChangeFormData} />
              {errors.subtitle && <small className="error">{errors.subtitle}</small>}

              <FormLabel title="Description" required length={250} currentLength={description.length} />
              <textarea className="mission-input mission-input--multi" type="text" name="description" placeholder="Here will be templated text but with lower opacity as lower opacity indicates that it is a placeholder. When user start to type within field, text should have 100% opacity." maxLength="250" value={description} onChange={this.handleChangeFormData} />
              {errors.description && <small className="error">{errors.description}</small>}

              <div className="mission-image-container">
                <div className="mission-image-container__half">
                  <FormLabel title="Mission Logo" required />
                  <img src={logoPreview} width={150} height={100} />
                  <input id="mission_logo" type="file" name="logo" onChange={this.handleImageChange} style={{display: 'none'}} />
                  <div className="mission-choose-file">
                    <label htmlFor="mission_logo" className="mission-input mission-input--file">
                      Choose file
                    </label>
                    <div className="img-resolution">Image should be <br />1200px x 800px</div>
                  </div>
                  {errors.logo && <small className="error">{errors.logo}</small>}
                </div>
                <div className="mission-image-container__half">
                  <FormLabel title="Mission Image" required />
                  <img src={imagePreview} width={100} height={100} />
                  <input id="mission_image" type="file" name="image" onChange={this.handleImageChange} style={{display: 'none'}} />
                  <div className="mission-choose-file">
                    <label htmlFor="mission_image" className="mission-input mission-input--file">
                      Choose file
                    </label>
                    <div className="img-resolution">Image should be <br />800px x 800px</div>
                  </div>
                  {errors.image && <small className="error">{errors.image}</small>}
                </div>
              </div>
            </div>
          </div>
        </div>

        <div className="mission-footer">
          <button className="mission-footer__save">{id ? 'Update' : 'Save'}</button>
        </div>
      </form>
    </React.Fragment>
  }
}

Mission.propTypes = {
  mission: PropTypes.shape({})
}
