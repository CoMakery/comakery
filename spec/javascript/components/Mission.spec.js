import React from 'react'
import { mount } from 'enzyme'
import Mission from 'components/Mission'

describe('Mission', () => {
  it('renders correctly without props', () => {
    const wrapper = mount(<Mission />)

    expect(wrapper).toMatchSnapshot()
    expect(wrapper.exists('.mission-container')).toBe(true)
    expect(wrapper.exists('.mission-header')).toBe(true)
    expect(wrapper.exists('.mission-details')).toBe(true)
    expect(wrapper.exists('.mission-leaders')).toBe(true)
    expect(wrapper.exists('.mission-stats')).toBe(true)
    expect(wrapper.exists('.mission-projects')).toBe(true)
  })

  it('renders correctly with props', () => {
    const props = {
      csrfToken: 'CtB6XLxhOE4xZ+SZ3IxMAPYTZ18b4FLjEQqrecPJVoKibXnW98WIASgvKbbqbcoHPF0EdhePqyQ8VfTL/dubKQ==',
      leaders  : [
        {
          count      : 4,
          firstName  : 'User1',
          id         : 1,
          imageUrl   : '/attachments/image',
          lastName   : 'Last1',
          projectName: 'test live call'
        },
        {
          count      : 2,
          firstName  : 'User2',
          id         : 2,
          imageUrl   : '/attachments/image',
          lastName   : 'Last2',
          projectName: 'test live call'
        },
        {
          count      : 3,
          firstName  : 'User3',
          id         : 3,
          imageUrl   : '/attachments/image',
          lastName   : 'Last3',
          projectName: 'test live call'
        }
      ],
      mission: {
        description: 'Holochain Description',
        id         : 1,
        imageUrl   : '/attachments/image.jpg',
        logoUrl    : '/attachments/logo.jpg',
        name       : 'HOLOCHAIN',
        subtitle   : 'Test Mission',
        stats      : {batches: 19, interests: 1, projects: 6, tasks: 369}
      },
      tokens: {
        tokenCount: 6,
        tokens    : [
          {
            contractUrl: 'https://testnet.qtum.org/token/test',
            count      : 2,
            id         : 1,
            logoUrl    : '/attachments/test logo',
            name       : 'QRC Token 1',
            projectName: 'test live call',
            symbol     : 'BIG'
          },
          {
            contractUrl: 'https://testnet.qtum.org/token/test',
            count      : 2,
            id         : 2,
            logoUrl    : '/attachments/test logo',
            name       : 'QRC Token 2',
            projectName: 'test live call',
            symbol     : 'BIG'
          },
          {
            contractUrl: 'https://testnet.qtum.org/token/test',
            count      : 2,
            id         : 3,
            logoUrl    : '/attachments/test logo',
            name       : 'QRC Token 3',
            projectName: 'test live call',
            symbol     : 'BIG'
          },
          {
            contractUrl: 'https://testnet.qtum.org/token/test',
            count      : 2,
            id         : 4,
            logoUrl    : '/attachments/test logo',
            name       : 'QRC Token 4',
            projectName: 'test live call',
            symbol     : 'BIG'
          }
        ]
      },
      projects: [
        {
          editable  : true,
          interested: false,
          stats     : {
            batches  : 3,
            tasks    : 0,
            interests: 0
          },
          projectData: {
            team           : [],
            teamSize       : 0,
            defaultImageUrl: 'defaultImg',
            description    : 'yes',
            id             : 60,
            imageUrl       : 'testimage',
            owner          : 'Test User',
            squareUrl      : 'test image',
            title          : 'test live call',
            teamLeader     : {
              firstName: 'Test user',
              id       : 1,
              imageUrl : '/attachments/test.jpg',
              lastName : 'Test',
              nickname : ''
            }
          },
          tokenData: {
            logoUrl: '/attachments/token image',
            name   : 'QRC Token'
          }
        },
        {
          editable  : true,
          interested: false,
          stats     : {
            batches  : 3,
            tasks    : 0,
            interests: 0
          },
          projectData: {
            team           : [],
            teamSize       : 0,
            defaultImageUrl: 'defaultImg',
            description    : 'yes',
            id             : 61,
            imageUrl       : 'testimage',
            owner          : 'Test User',
            squareUrl      : 'test image',
            title          : 'test live call',
            teamLeader     : {
              firstName: 'Test user',
              id       : 1,
              imageUrl : '/attachments/test.jpg',
              lastName : 'Test',
              nickname : ''
            }
          },
          tokenData: {
            logoUrl: '/attachments/token image2',
            name   : 'QRC Token2'
          }
        }
      ],
      newProjectUrl: '/projects/new?mission_id=test'

    }
    const wrapper = mount(<Mission {...props} />)

    expect(wrapper.find('.mission-details__name').text()).toBe('HOLOCHAIN')
    expect(wrapper.find('.mission-details__subtitle').text()).toBe('Test Mission')
    expect(wrapper.find('.mission-details__description').text()).toBe('Holochain Description')

    expect(wrapper.find('.mission-leaders__individual')).toHaveLength(3)
    expect(wrapper.contains(<div className="mission-leaders__individual__info__name">User1 Last1</div>)).toEqual(true)
    expect(wrapper.contains(<div className="mission-leaders__individual__info__name">User2 Last2</div>)).toEqual(true)
    expect(wrapper.contains(<div className="mission-leaders__individual__info__name">User3 Last3</div>)).toEqual(true)

    expect(wrapper.find('.mission-stats__kpi')).toHaveLength(3)
    expect(wrapper.find('.mission-projects__single')).toHaveLength(2)
  })
})
