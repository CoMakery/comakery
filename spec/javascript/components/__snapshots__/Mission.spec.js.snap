// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Mission renders correctly without props 1`] = `
<Mission
  leaders={Array []}
  mission={Object {}}
  projects={Array []}
  stats={Object {}}
  tokens={
    Object {
      "tokens": Array [],
    }
  }
>
  <div
    className="mission-container"
  >
    <div
      className="mission-header"
    >
      <div
        className="mission-header__blur"
        style={
          Object {
            "backgroundImage": "url(undefined)",
          }
        }
      />
      <div
        className="mission-header__logo"
      >
        <img
          className="mission-header__logo__img"
        />
      </div>
      <a
        className="mission-header__link"
      >
        CREATE A PROJECT
      </a>
    </div>
    <div
      className="mission-details"
    >
      <div
        className="mission-details__heading"
      >
        Mission
      </div>
      <div
        className="mission-details__name"
      />
      <div
        className="mission-details__subtitle"
      />
      <div
        className="mission-details__description"
      />
    </div>
    <div
      className="mission-leaders"
    >
      Featured Project Leaders
      <div
        className="mission-leaders__wrapper"
      >
        <div
          className="mission-leaders__own"
        >
          Want to be a project leader for this mission?
          <a
            className="mission-leaders__own__create"
          >
            CREATE YOUR OWN PROJECT
          </a>
        </div>
      </div>
    </div>
    <div
      className="mission-tokens"
    >
      Tokens Used In The Mission
      <div
        className="mission-tokens__cards"
      />
    </div>
    <div
      className="mission-stats"
    >
      Mission Stats
      <div
        className="mission-stats__container"
      >
        <div
          className="mission-stats__kpis"
        >
          <div
            className="mission-stats__kpis__row"
          >
            <div
              className="mission-stats__kpi"
            >
              <Icon
                className=""
                name="PROJECT/WHITE.svg"
              >
                <img
                  className="icon icon__PROJECT/WHITE-svg"
                  src="test-file-stub"
                />
              </Icon>
              <Plural
                className={null}
                count={1}
                showCount={true}
                singular="project"
                style={Object {}}
                zero={null}
              >
                <span
                  className={null}
                  style={Object {}}
                >
                  1 project
                </span>
              </Plural>
            </div>
            <div
              className="mission-stats__kpi"
            >
              <Icon
                className=""
                name="BATCH/WHITE.svg"
              >
                <img
                  className="icon icon__BATCH/WHITE-svg"
                  src="test-file-stub"
                />
              </Icon>
              <Plural
                className={null}
                count={1}
                plural="batches"
                showCount={true}
                singular="batch"
                style={Object {}}
                zero={null}
              >
                <span
                  className={null}
                  style={Object {}}
                >
                  1 batch
                </span>
              </Plural>
            </div>
          </div>
          <div
            className="mission-stats__kpis__row"
          >
            <div
              className="mission-stats__kpi"
            >
              <Icon
                className=""
                name="TASK/WHITE.svg"
              >
                <img
                  className="icon icon__TASK/WHITE-svg"
                  src="test-file-stub"
                />
              </Icon>
              <Plural
                className={null}
                count={1}
                showCount={true}
                singular="task"
                style={Object {}}
                zero={null}
              >
                <span
                  className={null}
                  style={Object {}}
                >
                  1 task
                </span>
              </Plural>
            </div>
            <div
              className="mission-stats__kpi"
            >
              <div
                className="mission-stats__contributor"
              />
              <Plural
                className={null}
                count={1}
                showCount={true}
                singular="person"
                style={Object {}}
                zero={null}
              >
                <span
                  className={null}
                  style={Object {}}
                >
                  1 person
                </span>
              </Plural>
               INTERESTED
            </div>
          </div>
          <div
            className="mission-stats__kpi mission-stats__kpi--desktop"
          >
            <Icon
              className=""
              name="PROJECT/WHITE.svg"
            >
              <img
                className="icon icon__PROJECT/WHITE-svg"
                src="test-file-stub"
              />
            </Icon>
            <Plural
              className={null}
              count={1}
              showCount={true}
              singular="project"
              style={Object {}}
              zero={null}
            >
              <span
                className={null}
                style={Object {}}
              >
                1 project
              </span>
            </Plural>
          </div>
          <div
            className="mission-stats__kpi mission-stats__kpi--desktop"
          >
            <Icon
              className=""
              name="BATCH/WHITE.svg"
            >
              <img
                className="icon icon__BATCH/WHITE-svg"
                src="test-file-stub"
              />
            </Icon>
            <Plural
              className={null}
              count={1}
              plural="batches"
              showCount={true}
              singular="batch"
              style={Object {}}
              zero={null}
            >
              <span
                className={null}
                style={Object {}}
              >
                1 batch
              </span>
            </Plural>
          </div>
          <div
            className="mission-stats__kpi mission-stats__kpi--desktop"
          >
            <Icon
              className=""
              name="TASK/WHITE.svg"
            >
              <img
                className="icon icon__TASK/WHITE-svg"
                src="test-file-stub"
              />
            </Icon>
            <Plural
              className={null}
              count={1}
              showCount={true}
              singular="task"
              style={Object {}}
              zero={null}
            >
              <span
                className={null}
                style={Object {}}
              >
                1 task
              </span>
            </Plural>
          </div>
          <div
            className="mission-stats__kpi mission-stats__kpi--desktop"
          >
            <div
              className="mission-stats__contributor"
            />
            <Plural
              className={null}
              count={1}
              showCount={true}
              singular="person"
              style={Object {}}
              zero={null}
            >
              <span
                className={null}
                style={Object {}}
              >
                1 person
              </span>
            </Plural>
             INTERESTED
          </div>
        </div>
      </div>
    </div>
    <div
      className="mission-projects"
    >
      <div
        className="mission-projects__container"
      >
        <div
          className="mission-projects__heading"
        >
          Featured Projects
        </div>
      </div>
    </div>
  </div>
</Mission>
`;
